### ISSUE TITLE
Merge getting duplicate key value violates unique constraint error

### ISSUE THREAD

**Issue by Snazzie** [zzzprojects/EntityFramework-Plus#739]
Here is what to include in your request to make sure we implement a solution as quickly as possible. 

## 1. Description
Im trying to insert data that may or may not already exist in the tables. Because i have a unique PK constraint, it fails when a row already exists.
I heard Merge should resolve this very problem as it does insert or update/Upsert. found out about it in [your comment on a stack overflow question](https://stackoverflow.com/a/36363651/4882318).
However, im still getting 23505: duplicate key value violates unique constraint from postgres.

## 2. Exception
If you are seeing an exception, include the full exception details (message and stack trace).

```
Exception message:
23505: duplicate key value violates unique constraint "13_13_PK_CpuMetrics"
DETAIL: Key ("TargetId", "Id", "TimeStamp")=(093a4cd8-6bc5-4a7a-a18f-063546c33e11, 005c1db4-fd16-4b38-bc78-ad78923a9242, 2022-04-10 01:54:04+00) already exists.
Stack trace:
   at Npgsql.Internal.NpgsqlConnector.<<ReadMessage>g__ReadMessageLong|211_0>d.MoveNext()
   at Npgsql.NpgsqlDataReader.<NextResult>d__47.MoveNext()
   at Npgsql.NpgsqlDataReader.NextResult()
   at Npgsql.NpgsqlCommand.<ExecuteReader>d__116.MoveNext()
   at Npgsql.NpgsqlCommand.<ExecuteReader>d__116.MoveNext()
   at Npgsql.NpgsqlCommand.<ExecuteNonQuery>d__104.MoveNext()
   at Npgsql.NpgsqlCommand.ExecuteNonQuery()
   at .(DbCommand , BulkOperation , Int32 )
   at .( , DbCommand )
   at .Execute(List`1 actions)
   at Z.BulkOperations.BulkOperation.Execute()
   at .BulkMerge[T](DbContext this, IEntityType entityType, IEnumerable`1 list, Action`1 options, SavingSelector savingSelector, Boolean forceSpecificTypeMapping)
   at .BulkMerge[T](DbContext this, IEnumerable`1 entities, Action`1 options, Boolean isBulkSaveChanges)
   at DbContextExtensions.BulkMerge[T](DbContext this, IEnumerable`1 entities, Action`1 options)
   at System.Threading.Tasks.Task.InnerInvoke()
   at System.Threading.Tasks.Task.<>c.<.cctor>b__272_0(Object obj)
   at System.Threading.ExecutionContext.RunFromThreadPoolDispatchLoop(Thread threadPoolThread, ExecutionContext executionContext, ContextCallback callback, Object state)
--- End of stack trace from previous location ---
   at System.Threading.ExecutionContext.RunFromThreadPoolDispatchLoop(Thread threadPoolThread, ExecutionContext executionContext, ContextCallback callback, Object state)
   at System.Threading.Tasks.Task.ExecuteWithThreadLocal(Task& currentTaskSlot, Thread threadPoolThread)
--- End of stack trace from previous location ---
   at DbContextExtensions.`1.MoveNext()
   at TimeSeriesShared.TimeScaleWriterClient.<InsertMetrics>d__7.MoveNext() in F:\Repo\Pluto\Source\MicroServices\TimeSeriesShared\TimeScaleWriterClient.cs:line 34
```

## 3. Fiddle or Project

```csharp
        public async Task<bool> InsertMetrics(IEnumerable<TimescaleBase> metrics)
        {
                var context = DbContext.CreateDbContext();

                var cat = metrics.GroupBy(e => e.GetType());

                foreach(var metric in cat)
                {
                    await context.BulkMergeAsync(metric, operation =>
                    {
                        operation.AllowDuplicateKeys = true;
                        operation.AllowUpdatePrimaryKeys = true;
                        operation.ColumnPrimaryKeyExpression =
                             m => m.Id;
                    });
                }
                return true;
        }
```
```csharp
    public abstract class TimescaleBase
    {
        [Required]
        [Column(TypeName = "timestamp with time zone")]
        public DateTime TimeStamp { get; set; } 

        [Required]
        [Column(TypeName = "TEXT")]
        public string TargetId { get; set; }

        [DatabaseGenerated(DatabaseGeneratedOption.Identity)]
        public Guid Id { get; set; } = new Guid();
    }
    
    public class CpuMetrics: TimescaleBase
    {
        public int CpuPercentage { get; set; }

        public int? CpuTemperature { get; set; }
    }
```
```csharp
    public class TimescaleDbContext : DbContext
    {
        public TimescaleDbContext(DbContextOptions<TimescaleDbContext> options) : base(options)
        {
        }
        public DbSet<CpuMetrics> CpuMetrics { get; set; }
        //...

        protected override void OnModelCreating(ModelBuilder modelBuilder)
        {
            modelBuilder.Entity<CpuMetrics>().HasKey(e => new { e.TargetId, e.Id, e.TimeStamp });
            //...
        }
    }
```
```sql
CREATE TABLE IF NOT EXISTS public."CpuMetrics"
(
    "CpuPercentage" integer NOT NULL DEFAULT 0,
    "CpuTemperature" integer,
    "TimeStamp" timestamp with time zone NOT NULL,
    "TargetId" text COLLATE pg_catalog."default" NOT NULL,
    "Id" uuid NOT NULL DEFAULT '00000000-0000-0000-0000-000000000000'::uuid,
    CONSTRAINT "PK_CpuMetrics" PRIMARY KEY ("TargetId", "Id", "TimeStamp")
)
```
## 4. Any further technical details
DB: Postgres
Package version: Z.EntityFramework.Plus.EFCore 6.14.4

**Comment by JonathanMagnan** [comment]
Hello @Snazzie ,

You have a unique constraint in your database on the following columns:
- TargetId
- Id
- TimeStamp

(Look like to be your key as well)

But you currently only use `Id` as you key.

So let say for the same `Id` you have multiple rows with different `TargetId`, but now, you want to update all of them with the same `TargetId` (as multiple rows match the `Id` key), so that means the `unique constraint` error will be raised.

Let me know if that explains correctly why this is happening and if you have enough information to solve it on your side since it looks like this issue is not coming from our library but more causing due to how you use the key options.

Best Regards,

Jon

**Comment by Snazzie** [comment]
To give more context, `Id` is always unique and is generated by db, `TargetId` and `Timestamp` is not unique. 
Rows in tables would never contain the same Id, but it can contain same targetId and timestamp, there may be additional columns to differentiate the metric further. 
For example if im collecting disk data and there are 2 disks on a target machine, there will be a text column for Disk Letter, the two metrics entries still belongs to the same targetIds and time of metric recorded is the same.

```
metrics... | letter | targetId    | Id                      | timestamp
------------------------------------------------------------------------------
metrics... | C:     | 1           | 1234567-13112-1237-1231 | 2022-01-01 12:00:00
metrics... | D:     | 1           | 8910111-13123-1412-1242 | 2022-01-01 12:00:00
```
The reason for this way of doing is that timescaleDb does not allow creations of "hyper tables" on tables that contain primary keys, and without keys EF Core does not work. The only way to satisfy both is to create a composite key.

Unfortunately my knowledge in SQL is quite minimal and cant immediately see what can be done here, I did try providing all three properties into `ColumnPrimaryKeyExpression` but obviously that wasnt allowed.

**Comment by Snazzie** [comment]
Based on the above explanation, i feel like i shouldnt be getting a unique constraint error? isnt the whole point is to update instead of insert if data already exists?

**Comment by Snazzie** [comment]
think im just a bit daft.

i tried ` operation.ColumnPrimaryKeyExpression = m=> new { m.TargetId, m.Id, m.TimeStamp };` and it seems to be ok with it now.  before i was trying to `operation.ColumnPrimaryKeyExpression =  m => (m.TargetId, m.Id, m.TimeStamp)` 🤦

**Comment by Snazzie** [comment]
just to mention i tried `operation.ColumnPrimaryKeyExpression = metric => new { metric.TargetId, metric.Id, metric.TimeStamp };` as i thought that could fix it, didnt resolve it :/

**Comment by Snazzie** [comment]
Im going to do a proper investigation, i made a bunch of changes and it just suddenly worked and i dont know what exactly i did in the end that fixed it.

**Comment by JonathanMagnan** [comment]
Hello @Snazzie ,

Do you think you could create a runnable project with the issue? It doesn’t need to be your project, just a new solution with the minimum code to reproduce the issue.

I don't think without the data, we can really reproduce it.

Your key should only be the `Id` column as this one is unique. Adding more columns to the key has no value. Some additional constraint indexes could exist, but the key is what makes a row unique (the `Id` as you specified).

Maybe we are missing something important here, so having a project will help us discover why exactly this is happening.

Best Regards,

Jon

**Comment by Snazzie** [comment]
@JonathanMagnan 

What's interesting is that, the `Id`s should have never clashed, as all objects in `metrics` had `Id` set to empty guid. 

So i did manage to get it working. But i didn't figure out the exact cause of the problem. I did catch a bug in my code that was spawning multiple threads trying to merge the same set of data at the same time on seperate connection, which may or may not contribute to the above? 

I created a demo proj, tried a few different hypotheses and they all passed without a problem.
I tried running them in parallel and that passed as well, although im only using 2 entries vs 5000.

I'll close this as it seems like Merge is working correctly as is and am unable to provide a reproducible failure and dont wanna spend more time on it 😅.
